components/family-studies/api/src/main/java/org/phenotips/studies/family/internal/DenyAccessByDefaultAuthorizationModule.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/
+ */
+package org.phenotips.studies.family.internal;
+
+import org.phenotips.security.authorization.AuthorizationModule;
+import org.phenotips.studies.family.Family;
+import org.phenotips.studies.family.FamilyRepository;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.EntityType;
+import org.xwiki.model.reference.EntityReference;
+import org.xwiki.security.authorization.Right;
+import org.xwiki.users.User;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.ObjectUtils;
+
+/**
+ * Low priority implementation that by default denies access to family records.
+ *
+ * @version $Id$
+ * @since 1.4
+ */
+@Component
+@Named("family-deny-access")
+@Singleton
+public class DenyAccessByDefaultAuthorizationModule implements AuthorizationModule
+{
+    /** Checks to see if a document is a family. */
+    @Inject
+    private FamilyRepository familyRepository;
+
+    @Override
+    public int getPriority()
+    {
+        return 110;
+    }
+
+    @Override
+    public Boolean hasAccess(User user, Right access, EntityReference entity)
+    {
+        if (!ObjectUtils.allNotNull(access, entity) || access.getTargetedEntityType() == null
+            || !access.getTargetedEntityType().contains(EntityType.DOCUMENT)) {
+            return null;
```

Comment:
Just curious what is the point of doing any checks and returning `null` instead of just always returning `false` (presumably "deny by default" should always return `false`?)

Reply:
* `ObjectUtils.allNotNull(access, entity)` - sanity check
* `access.getTargetedEntityType() == null || || !access.getTargetedEntityType().contains(EntityType.DOCUMENT)` - XWiki supports rights that are not specific to a document, e.g. administration, registration, scripting, programming. It is not the responsibility of this module to restrict those rights.

Reply:
I see. I understood the sanity check - but just thought that even in "insane" conditions the "deny" module should deny

("not the responsibility of this module to restrict those right" - aha, this I did not know and one of the reasons I asked)

Reply:
BTW, which module should be responsible for that? Can someone please describe the way the entire permission system works now?

Reply:
These are instance-level rights. They do not concern documents, but actions such accessing `/register/` for creating accounts, or `/admin/` for administering. They are the responsibility of the XWiki platform authorization module, which is the last one interrogated when we return null here.

Reply:
I see, and I presume it will return `false` in all the insane cases

Reply:
yes

Reply:
> just thought that even in "insane" conditions the "deny" module should deny
This is the deny module _for families_, so it should not deny everything, just access to family documents. But if the pieces required to determine if this is a request to access a document and of that document is a family are missing, this module can not (and should not) decide anything.

Even after the [XWikiACL](https://github.com/phenotips/phenotips/blob/master/components/security/authorization/src/main/java/org/phenotips/security/authorization/internal/XWikiACLAuthorizationModule.java) module runs, there is [https://github.com/phenotips/phenotips/blob/master/components/security/authorization/src/main/java/org/phenotips/security/authorization/internal/BaseAuthorizationModule.java](another failsafe module) that returns false unconditionally.

Reply:
right